crowbar_framework/app/models/crowbar_service.rb

```diff
@@ -169,6 +182,63 @@ def create_proposal
     base
   end
 
+  def prepare_nodes_for_crowbar_upgrade
+    proposal = ProposalObject.find_proposal('crowbar', 'default')
+
+    # add all nodes new role which prepares them for the upgrade
+    nodes_to_upgrade = []
+    NodeObject.all.each do |node|
+      if node.state == "ready" && !node.admin?
```

Comment:
Oh, actually: the `node.state == "ready"` bit feels wrong to me. If a node is not ready, then we should simply block the upgrade (ie, fail here). The only exceptions are if a node is discovered/discovering.